File: 20230807/python/hyeonjeong.py
import sys
sys.setrecursionlimit(10**6)

def solution(n, m, x, y, r, c, k):
    
    dir = {'d':[1, 0], 'l':[0, -1], 'r':[0, 1], 'u':[-1, 0]}
    answer = ""
    
    def dfs(x, y, temp):
        nonlocal answer
        # 이동 거리가 k이고 위치가 탈출 지점이면 경로를 저장하고 나가기
        if len(temp) == k and x == r and y == c:
            answer = temp
            return
        # 지금까지 온 거리와 남은 거리가 정해진 k 이동 거리를 초과하면 나가기
        if abs(r-x) + abs(c-y) + len(temp) > k:
            return
        # 상하좌우 돌면서
        for letter, pos in dir.items():
            # 바깥으로 안나가면서 아직 정답이 안정해졌으면 이동하기
            if 1 <= x + pos[0] <= n and 1 <= y + pos[1] <= m and len(answer) == 0:
                dfs(x + pos[0], y + pos[1], temp + letter)
    
    # 최단 이동거리가 k를 초과하거나 k번 째에 탈출 지점으로 이동할 수 없으면 실패
    dist = abs(r-x) + abs(c-y)
    if dist > k or (k-dist) % 2 != 0:
        return "impossible"
    
    dfs(x, y, "")
    
    return answer
    
    

File: 20231008/java/hyeonjeong.java
class Solution {
    int MOD = 20170805;
    int[][] directions = {{0, 1}, {1, 0}};
    int m, n, answer;
    int[][] cityMap;
    
    public int solution(int m, int n, int[][] cityMap) {
        
        dfs(0, 0, 0);
        dfs(0, 0, 1);
        
        return answer % MOD;
        
    }
    public void dfs(int i, int j, int dir) {
        if (i==m-1 && j==n-1) {
            answer++;
            return;
        }

        int ni = i + directions[dir][0];
        int nj = j + directions[dir][1];
        if (ni < 0 || ni >= m || nj < 0 || nj >= n) return;
        if (cityMap[ni][nj] != 2) dfs(ni, nj, dir);
        else if (cityMap[ni][nj] == 0) {
            dfs(ni, nj, 0);
            dfs(ni, nj, 1);
        }
    
    }
}
